torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -234,8 +236,64 @@ inline void errorIfCapturingNonCapturableNCCL() {
   }
 }
 
+void markRankNotStuck(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x00;
+}
+
+void markRankStuck(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x01;
+}
+
+void markRankTimedOut(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x02;
+}
+
 } // namespace
 
+RankStatusHelper::RankStatusHelper(size_t worldSize): worldSize_(worldSize) {
+  // assume file exists
+  int fd;
+  if ((fd = open("rank_statuses", O_RDWR, 0)) == -1) {
+    LOG(ERROR) << "unable to open rank_statuses";
+  }
+  handle_ = static_cast<char*>(mmap(
+                        nullptr,
+                        worldSize,
+                        PROT_WRITE | PROT_READ,
+                        MAP_SHARED,
+                        fd,
+                        /* offset */ 0));
+}
+
+RankStatusHelper::~RankStatusHelper() {
+  munmap(handle_, worldSize_);
```

Comment:
Don't we need to do an `msync` to flush all the data before we unmap? My understanding is that MAP_SHARED would write the data back to the file asynchronously so there is no guarantee it would be completed unless we do msync?

Reply:
> msync()  flushes changes made to the in-core copy of a file that was mapped into memory using mmap(2) back to the filesystem.  Without use of this call, there is no guarantee that changes are written back before munmap(2) is called.

This makes me believe that calling `munmap` should guarantee that the changes are written back to the filesystem

Reply:
Is there a reason we don't want to delete + process the file here? I'm guessing for efficiency reasons it won't be feasible, as we only want to do it once when the job is killed/stuck/terminating?